src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -551,42 +551,46 @@ private async Task OnBatchScopeDisposedMaybeAsync(bool useAsync)
+            // The transformation function will set these variables, but we need to use them after the transformation is applied
+            // so we must instantiate them here.
+            ImmutableArray<string> documentFileNamesAdded = [];
+            ImmutableArray<(DocumentId documentId, SourceTextContainer textContainer)> documentsToOpen = [];
+            ImmutableArray<(DocumentId documentId, SourceTextContainer textContainer)> additionalDocumentsToOpen = [];
+            ImmutableArray<(DocumentId documentId, SourceTextContainer textContainer)> analyzerConfigDocumentsToOpen = [];
 
             var hasAnalyzerChanges = _analyzersAddedInBatch.Count > 0 || _analyzersRemovedInBatch.Count > 0;
 
             await _projectSystemProjectFactory.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync, (solutionChanges, projectUpdateState) =>
             {
-                _sourceFiles.UpdateSolutionForBatch(
+                // Changes made inside this transformation must be idemopotent in case it is attempted multiple times.
+
+                var documentFileNamesAddedBuilder = ImmutableArray.CreateBuilder<string>();
+                documentsToOpen = _sourceFiles.UpdateSolutionForBatch(
                     solutionChanges,
-                    documentFileNamesAdded,
-                    documentsToOpen,
+                    documentFileNamesAddedBuilder,
                     static (s, documents) => s.AddDocuments(documents),
                     WorkspaceChangeKind.DocumentAdded,
                     static (s, ids) => s.RemoveDocuments(ids),
                     WorkspaceChangeKind.DocumentRemoved);
 
-                _additionalFiles.UpdateSolutionForBatch(
+                additionalDocumentsToOpen = _additionalFiles.UpdateSolutionForBatch(
                     solutionChanges,
-                    documentFileNamesAdded,
-                    additionalDocumentsToOpen,
+                    documentFileNamesAddedBuilder,
                     static (s, documents) => s.AddAdditionalDocuments(documents),
                     WorkspaceChangeKind.AdditionalDocumentAdded,
                     static (s, ids) => s.RemoveAdditionalDocuments(ids),
                     WorkspaceChangeKind.AdditionalDocumentRemoved);
 
-                _analyzerConfigFiles.UpdateSolutionForBatch(
+                analyzerConfigDocumentsToOpen = _analyzerConfigFiles.UpdateSolutionForBatch(
                     solutionChanges,
-                    documentFileNamesAdded,
-                    analyzerConfigDocumentsToOpen,
+                    documentFileNamesAddedBuilder,
                     static (s, documents) => s.AddAnalyzerConfigDocuments(documents),
                     WorkspaceChangeKind.AnalyzerConfigDocumentAdded,
                     static (s, ids) => s.RemoveAnalyzerConfigDocuments(ids),
                     WorkspaceChangeKind.AnalyzerConfigDocumentRemoved);
```

Comment:
i'm missing something.  who actually reads these xxxDocumentToOpen after you've assigned into them.  i don't see reads later on.

Reply:
talked offline, this is read on line 707